File: c++/src/kmc_functions.cpp
#ifndef PLUGIN_BACKENDS_
#include "kmc_functions.h"
#endif

std::vector<double> collect_coverages(const std::vector<std::string> & types,
                                      const std::vector<std::string> & possible_types,
                                      const std::vector<double> & coverage_ratios)
{
    // Get ntotal umber of sites.
    const size_t nsite = types.size()/coverage_ratios.size();

    // Number of different types.
    std::vector<double> ntypes(possible_types.size(), 0.0);

    // Variables for coverage ratio.
    double ratio;
    int ratio_idx;

    // Loop over all types to collect coverages.
    for (size_t type_idx = 0; type_idx < types.size(); ++type_idx)
    {
        // Current element.
        const std::string & element = types[type_idx];

        // Get corresponding coverage ratio.
        ratio_idx = static_cast<int>(type_idx % coverage_ratios.size());
        ratio = coverage_ratios[ratio_idx];

        // Find the element in possible types, accumulate coverage.
        DoubleIterType ntypes_it = ntypes.begin();
        ConstStrIterType possible_types_it = possible_types.begin();
        const ConstStrIterType end = possible_types.end();

        for (; possible_types_it != end; ++ntypes_it, ++possible_types_it)
        {
            if (element == *possible_types_it)
            {
                // Accumulate.
                *ntypes_it += ratio;
            }
        }
    }

    // Calculate coverages.
    std::vector<double> coverages;
    double coverage;

    for (double ntype : ntypes)
    {
        coverage = ntype/static_cast<double>(nsite);
        coverages.push_back(coverage);
    }

    return coverages;
}

File: scaks/solvers/kmc_plugins/kmc_functions.py
import numpy as np


def collect_coverages(types, possible_types, coverage_ratios):
    """
    Function to get current coverages of possible types.

    :param types: The site types at the lattice points as a list
    :type types: list of str

    :param possible_types: possible species type in grid
    :type possible_types: list of str

    :param coverage_ratios: The coverages ratios for all basis sites
    :type coverage_ratios: list of float

    :return: coverages of possible types
    :rtype: numpy.array of int
    """
    # Total number of sites.
    nsite = len(types)//len(coverage_ratios)

    # Numbers of different types.
    ntypes = [0]*len(possible_types)

    # Loop to collect coverages.
    all_ratios = coverage_ratios*nsite
    for ratio, element in zip(all_ratios, types):
        # Ignore the empty type.
        if element not in possible_types:
            continue

        idx = possible_types.index(element)
        ntypes[idx] += ratio

    cvgs = [float(ntype)/nsite for ntype in ntypes]

    return cvgs
